core/services/run_queue.go

```diff
@@ -34,81 +36,188 @@ type RunQueue interface {
-		return
+	select {
+	case rq.chRuns <- run:
+	case <-rq.chStop:
 	}
+}
 
-	runID := run.ID.String()
-	defer numberRunsQueued.Inc()
-	if queueCount, present := rq.workers[runID]; present {
-		rq.workers[runID] = queueCount + 1
-		rq.workersMutex.Unlock()
-		return
-	}
-	rq.workers[runID] = 1
-	numberRunQueueWorkers.Set(float64(len(rq.workers)))
-	rq.workersMutex.Unlock()
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.JobSpecID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.JobSpecID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+				worker.chAddRun <- run
+			} else {
+				select {
+				case worker.chAddRun <- run:
+				case <-worker.chDone:
+					// If the worker is just spinning down as new work is coming in, allow it
+					// to die and spin up a new one.
+					worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+					rq.workers[*run.JobSpecID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+					worker.chAddRun <- run
+				}
+			}
 
-	rq.workersWg.Add(1)
-	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
+		case jobSpecID := <-rq.chWorkerDone:
+			delete(rq.workers, jobSpecID)
+			if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+				panic("numWorkers should never be < 0")
 			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+		case <-rq.chStop:
+			for _, worker := range rq.workers {
+				worker.chStop <- struct{}{}
+				<-worker.chDone
+				if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+					panic("numWorkers should never be < 0")
+				}
 			}
+			return
 		}
+	}
+}
 
-		rq.workersWg.Done()
+func (rq *runQueue) runSingleJobSpecWorker(jobSpecID models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
+
+	promNumberRunQueueWorkersStarted.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan []models.ID)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
+	go func() {
+		for runsRequested := range chResume {
```

Comment:
`runsRequested` is an array of runIDs? I find this confusing that it's a channel of arrays? Why not pass individual IDs? They're still serialized in the for loop, right?

Reply:
I don't think there's any way to do this without either 1) spinning up yet another goroutine to send queued runs one-by-one to the worker, or 2) blocking the coordinator while a job is running.  Neither of those are desirable.